app/controllers/application_controller.rb

```diff
@@ -25,8 +25,8 @@ def append_info_to_payload(payload)
     payload[:ip] = request.remote_ip
   end
 
-  def analytics
-    @analytics ||= Analytics.new(current_user, request_attributes, ahoy)
+  def analytics(service = Analytics.new(current_user, request))
+    @analytics ||= service
```

Comment:
The instance variable is cached per request. If one controller method calls `analytics`, and then another controller method calls `analytics(FakeAnalytics.new)`, the regular `Analytics` class will be called in the first controller, and the fake will be called in the second controller. In practice, the fake will only be called in tests.